docs/technical/git-submodules.md

```diff
@@ -0,0 +1,167 @@
+is sometimes called the "absorbed form", and is Git's preferred mode of
+operation.
+
+## Roadmap
+
+Git submodules should be implemented in an order that supports an increasing set
+of workflows, with the goal of getting feedback early and often. When support is
+incomplete, jj should not crash, but instead provide fallback behavior and warn
+the user where needed.
+
+The goal is to land good support for pure Jujutsu repositories, while colocated
+repositories will be supported when convenient.
+
+This section should be treated as a set of guidelines, not a strict order of
+work.
+
+### Phase 1: Readonly submodules
+
+This includes work that inspects submodule contents but does not create new
+objects in the submodule. This requires a way to store submodules in a jj
+repository that supports readonly operations.
+
+#### Outcomes
+
+- Submodules can be cloned anew
+- New submodule commits can be fetched
+- Submodule history and branches can be viewed
+- Submodule contents are populated in the working copy
+- Superproject gitlink can be updated to an existing submodule commit
+- Conflicts in the superproject gitlink can be resolved to an existing submodule
+  commit
+
+### Phase 2: Snapshotting new changes
+
+This allows a user to write new contents to a submodule and its remote.
+
+#### Outcomes
+
+- Changes in the working copy can be recorded in a submodule commit
+- Submodule branches can be modified
+- Submodules and their branches can be pushed to their remote
+
+### Phase 3: Merging/rebasing/conflicts
+
+This allows merging and rebasing of superproject commits in a content-aware way
+(in contrast to Git, where only the gitlink commit ids are compared), as well as
+workflows that make resolving conflicts easy and sensible.
+
+This can be done in tandem with Phase 2, but will likely require a significant
+amount of design work on its own.
+
+#### Outcomes
+
+- Merged/rebased submodules result in merged/rebased working copy content
+- Merged/rebased working copy content can be committed, possibly by creating
+  sensible merged/rebased submodule commits
+- Merge/rebase between submodule and non-submodule gives a sensible result
+- Merge/rebase between submodule A and submodule B gives a sensible result
+
+### Phase ?: An ideal world
+
+I.e. outcomes we would like to see if there were no constraints whatsoever.
+
+- Rewriting submodule commits rewrites descendants correctly and updates
+  superproject gitlinks.
+- Submodule conflicts automatically resolve to the 'correct' submodule commits,
+  e.g. a merge between superproject commits creating a merge of the submodule
+  commits.
+- Nested submodules are as easy to work with as non-nested submodules.
+
+## Design
+
+### Guiding principles
+
+TODO
+
+### On-disk representation
+
+In the Git backend, submodules will be stored in Git's "absorbed submodule"
+format, i.e. a full Git repository with a Git directory in
+`<superproject-git-dir>/modules/<submodule-name>`, and a working tree with a
```

Comment:
maybe you can already establish that <name> is % encoded to avoid the bifurcation that Google's git deployment has here.

Reply:
I'm sorry, I was confused about the general gist of your idea. Git's tendency to put things called `$NAME` into `.git/somedir/$NAME` is a bad idea for the following reasons:

* security (what if $NAME = "../../blah")
* portability (what about case-insensitive vs case-sensitive filesystems?)
* functionality (dir/file conflicts)

I was hoping that JJ's approach wouldn't make the same mistake here, and could specify something that always escapes the name.

Reply:
Ah I see, I misunderstood. You are suggesting that `jj` should preemptively %-encode the name, using Google's Git patches as a example of what _not_ to do. I think this makes sense. Doing this means that in colocated repos, Git won't understand the submodule directory name, but we can upstream the change into Git. (That change will probably be welcome by the rest of the Google Git team, since we can drop the downstream patch.)